### PR TITLE
Allocate one cost tracker to regulate forwardable batches

### PR DESCRIPTION
**Pull request by tao-stones**
#### Problem

Forwarder has many ForwardBatches to cache forwardable packets; each ForwardBatch has its own instance of CostTrack, which has a HashMap. Every time `handle_forward` is call, many Hashmap are allocated and then deallocated, resulting  inefficient memory operations. 

ForwardBatches can share one cost tracker to significantly reduce hashmap (de)allocation. 

#### Summary of Changes
- Allocate one cost_tracker to regulate forwarding CU limits, instead of one per ForwardBatch
- Test new function

Fixes #

### CODE REVIEW

**Comment by tao-stones** [comment]
bench compare: 
master: 
```
test bench_forwarder_handle_forwading_contentious_transaction ... bench:  66,544,694 ns/iter (+/- 746,252)
test bench_forwarder_handle_forwading_parallel_transactions   ... bench:  24,283,999 ns/iter (+/- 380,415)
```

PR:
```
test bench_forwarder_handle_forwading_contentious_transaction ... bench:  13,766,795 ns/iter (+/- 195,305)
test bench_forwarder_handle_forwading_parallel_transactions   ... bench:  14,065,119 ns/iter (+/- 226,439)
```

About 5x improvement in `contentious` test, where more cost-trackers are created and destroyed. In `parallel` test, where much less cost-trackers are created, it still see about 2x improvement.

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs [hunk 1]
```diff
@@ -104,39 +74,86 @@ impl ForwardPacketBatchesByAccounts {
...
-                return true;
+        let tx_cost = CostModel::calculate_cost(sanitized_transaction, feature_set);
+
+        if let Ok(updated_costs) = self.cost_tracker.try_add(&tx_cost) {
+            let batch_index = self.get_batch_index_by_updated_costs(&tx_cost, &updated_costs);
+
+            if let Some(forward_batch) = self.forward_batches.get_mut(batch_index) {
+                forward_batch.forwardable_packets.push(immutable_packet);
+            } else {
+                // A successfully added tx_cost means it does not exceed block limit, nor vote
```
Comment (alessandrod) [inline core/src/banking_stage/forward_packet_batches_by_accounts.rs:115]: get_batch_index_by_updated_costs doesn't seem to mod?

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs [hunk 2]
```diff
@@ -333,4 +322,122 @@ mod tests {
...
+                        updated_costliest_account_cost: 0
+                    }
+                )
+            );
+            assert_eq!(
+                1,
+                forward_packet_batches_by_accounts.get_batch_index_by_updated_costs(
+                    &transaction_cost,
+                    &UpdatedCosts {
+                        updated_block_cost: test_cost * 2,
```
Comment (alessandrod) [inline core/src/banking_stage/forward_packet_batches_by_accounts.rs]: nit: this should probably be test_cost + 1 to test the boundary condition?

**Review by alessandrod — COMMENTED** [review]
Amazing thanks so much for doing this change! Lgtm just a couple of nits and that question about mod

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs (hunk 1 above)
Comment (tao-stones) [reply core/src/banking_stage/forward_packet_batches_by_accounts.rs:115]: Updated the comment - mistakenly thought to use "reminder" at first, then changed to divide but forgot update the comment. Thanks for pointing it out.

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs (hunk 2 above)
Comment (tao-stones) [reply core/src/banking_stage/forward_packet_batches_by_accounts.rs]: 100%, updated!

**Review by tao-stones — COMMENTED** [review]
thanks for reviewing, updated at 22d9cecb01aa070c389c76ff51795e74a7c9e239

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs [hunk 3]
```diff
@@ -104,39 +74,86 @@ impl ForwardPacketBatchesByAccounts {
...
+    // would be exceeded. Eg, if by block limit, it can be put into batch #1; by vote limit, it can
+    // be put into batch #2; and by account limit, it can be put into batch #3; then it should be
+    // put into batch #3 to satisfy all batch limits.
+    fn get_batch_index_by_updated_costs(
+        &self,
+        tx_cost: &TransactionCost,
+        updated_costs: &UpdatedCosts,
+    ) -> usize {
+        let batch_index_by_block_limit = updated_costs.updated_block_cost
+            / match tx_cost {
```
Comment (alessandrod) [inline core/src/banking_stage/forward_packet_batches_by_accounts.rs:140]: the divisors in this function can effectively never be 0 right?

**Review by alessandrod — APPROVED** [review]

File: core/src/banking_stage/forward_packet_batches_by_accounts.rs (hunk 3 above)
Comment (tao-stones) [reply core/src/banking_stage/forward_packet_batches_by_accounts.rs:140]: no, those limits must be above zero, otherwise nothing gets into block. But in any case, will change to checked_div in a follow up PR, so to be correct 😄
